In [27]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_absolute_percentage_error

from sklearn.ensemble import AdaBoostRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import precision_score

In [28]:
df = pd.read_csv('wk11_moreinfo.csv')

In [29]:
df

Wk      Time            Winner  Pts             Loser  Pts.1  \
0     1   7:00 PM  Florida Atlantic   43         Charlotte     13   
1     1   5:00 PM     Florida State   47          Duquesne      7   
2     1   4:00 PM          Illinois   38           Wyoming      6   
3     1  10:00 PM            Nevada   23  New Mexico State     12   
4     1   3:30 PM  Nevada-Las Vegas   52       Idaho State     21   
..   ..       ...               ...  ...               ...    ...   
644  11   7:30 PM         (12) Utah   45           Arizona     20   
645  11   3:30 PM        Utah State   27        New Mexico     10   
646  11   3:30 PM  Washington State   52          Stanford     14   
647  11  12:00 PM  Western Kentucky   59         Charlotte      7   
648  11  12:00 PM         Wisconsin   23          Maryland     10   

              winners            losers  home_rank  away_rank  home_team_code  \
0    Florida Atlantic         Charlotte         99         99              32   
1       Florida State          Duquesne         99         99              34   
2            Illinois           Wyoming         99         99              42   
3              Nevada  New Mexico State         99         99              76   
4    Nevada-Las Vegas       Idaho State         99         99              74   
..                ...               ...        ...        ...             ...   
644              Utah           Arizona         99         99             119   
645        Utah State        New Mexico         99         99             120   
646  Washington State          Stanford         99         99             103   
647  Western Kentucky         Charlotte         99         99              21   
648         Wisconsin          Maryland         99         99             130   

     away_team_code  home_results  win_rank  lose_rank       dates  day_code  
0                32             0        99         99  2022-08-27         5  
1                44             0        99         99  2022-08-27         5  
2               206             0        99         99  2022-08-27         5  
3               113             1        99         99  2022-08-27         5  
4                70             0        99         99  2022-08-27         5  
..              ...           ...       ...        ...         ...       ...  
644             207             0        99         99  2022-11-05         5  
645             116             0        99         99  2022-11-05         5  
646             196             1        99         99  2022-11-05         5  
647             201             1        99         99  2022-11-05         5  
648              95             0        99         99  2022-11-05         5  

[649 rows x 17 columns]

In [30]:
df = df.drop(columns=['Time', 'Winner', 'Loser', 'dates'])

In [32]:
df[df['winners'] == 'Alabama']

Wk  Pts  Pts.1  winners             losers  home_rank  away_rank  \
36    2   55      0  Alabama         Utah State         99         99   
97    3   20     19  Alabama              Texas         99          1   
178   4   63      7  Alabama   Louisiana-Monroe         99         99   
254   5   55      3  Alabama         Vanderbilt         99         99   
322   6   49     26  Alabama           Arkansas         99         99   
382   7   24     20  Alabama          Texas A&M         99         99   
492   9   30      6  Alabama  Mississippi State         99         99   

     home_team_code  away_team_code  home_results  win_rank  lose_rank  \
36                2             207             0         1         99   
97              107               3             1         1         99   
178               2             207             0         2         99   
254               2             207             0         2         99   
322               7             207             1         2         99   
382               2             207             0         1         99   
492               2             207             0         6         99   

     day_code  
36          5  
97          5  
178         5  
254         5  
322         5  
382         5  
492         5

In [33]:
df[df['winners'] == 'Mississippi']

Wk  Pts  Pts.1      winners            losers  home_rank  away_rank  \
62    2   28     10  Mississippi              Troy         99         99   
136   3   59      3  Mississippi  Central Arkansas         99         99   
213   4   42      0  Mississippi      Georgia Tech         99         99   
281   5   35     27  Mississippi             Tulsa         99         99   
352   6   22     19  Mississippi          Kentucky         99         99   
405   7   52     28  Mississippi        Vanderbilt         99          9   
463   8   48     34  Mississippi            Auburn         99         99   
570  10   31     28  Mississippi         Texas A&M         99         99   

     home_team_code  away_team_code  home_results  win_rank  lose_rank  \
62               67             207             0        99         99   
136              67             207             0        99         99   
213              39             106             1        99         99   
281              67             207             0        99         99   
352              67             207             0        99          7   
405             121             106             1         9         99   
463              67             207             0         9         99   
570             108             106             1        99         99   

     day_code  
62          5  
136         5  
213         5  
281         5  
352         5  
405         5  
463         5  
570         5

In [34]:
df['away_team_code'].value_counts()

207    140
24       6
32       5
66       5
151      5
      ... 
200      1
166      1
159      1
72       1
152      1
Name: away_team_code, Length: 208, dtype: int64

In [35]:
df['home_team_code'].value_counts()

72    11
4      6
24     6
53     6
43     6
      ..
55     4
62     3
99     3
65     3
17     3
Name: home_team_code, Length: 132, dtype: int64

In [6]:
# df = df.drop(columns=['Pts.1'])

In [7]:
y = df['Pts.1']
X = df.drop(columns=['Pts.1'])

X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.30, random_state=2)

In [8]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.datasets import make_regression

In [9]:
regr = RandomForestRegressor(max_depth=500, random_state=0)
regr.fit(X_train, y_train)

RandomForestRegressor(max_depth=500, random_state=0)

In [10]:
test_pred = regr.predict(X_test)

print('RMSE: '+str(np.sqrt(mean_squared_error(y_test, test_pred))))
print('R Squared: '+str(r2_score(y_test, test_pred)))
print('MSE TEST: '+str(mean_squared_error(y_test, test_pred)))
print('MAE TEST: '+str(mean_absolute_error(y_test, test_pred)))
print('MAPE: '+str(mean_absolute_percentage_error(y_test, test_pred)))

RMSE: 10.359420412457991
R Squared: 0.07023123090109895
MSE TEST: 107.31759128205128
MAE TEST: 7.624153846153846
MAPE: 3240744101090400.5


In [11]:
# MAPE: 0.3155694978397048

In [12]:
data = {'y_test': y_test, 'y_pred': test_pred}
new_df = pd.DataFrame(data)
diff = (new_df['y_test'] - new_df['y_pred']).abs()
new_df['diff'] = diff
new_df = new_df.sort_values(by='diff', ascending=True)
new_df

y_test  y_pred   diff
52       10   10.01   0.01
141      12   12.01   0.01
333      14   13.92   0.08
576      24   23.87   0.13
265      23   23.13   0.13
..      ...     ...    ...
569       0   24.11  24.11
108      57   19.06  37.94
20       44    5.65  38.35
288      52   12.86  39.14
67       61   18.22  42.78

[195 rows x 3 columns]

In [13]:
predvals = new_df['y_pred']

In [14]:
max_num = 0

for i in predvals:
    if (i > max_num):
        max_num = i
    else: 
        continue;

In [15]:
min_num = 100

for i in predvals:
    if (i < max_num):
        min_num = i
    else: 
        continue;

In [16]:
print(min_num)
print(max_num)

18.22
31.37


A MAPE less than 5% is considered as an indication that the forecast is acceptably accurate. A MAPE greater than 10% but less than 25% indicates low, but acceptable accuracy and MAPE greater than 25% very low accuracy, so low that the forecast is not acceptable in terms of its accuracy.

In [17]:
df.head()

Wk  Pts  Pts.1  home_rank  away_rank  home_team_code  away_team_code  \
0   1   43     13         99         99             136             117   
1   1   47      7         99         99             138             128   
2   1   38      6         99         99             145             289   
3   1   23     12         99         99             176             197   
4   1   52     21         99         99             174             154   

   home_results  win_rank  lose_rank  day_code  
0             0        99         99         5  
1             0        99         99         5  
2             0        99         99         5  
3             1        99         99         5  
4             0        99         99         5

In [18]:
# la = [[12, 40, 99, 7, 112, 174, 0, 99, 7, 5]]
# lsu = pd.DataFrame(la, columns=['Wk', 'Pts', 'home_rank', 'away_rank', 
#                                 'home_team_code', 'away_team_code', 
#                                 'home_results', 'win_rank', 'lose_rank', 'day_code'])
# lsu

la = [[12, 34, 99, 7, 112, 174, 1, 7, 99, 5]]
lsu = pd.DataFrame(la, columns=['Wk', 'Pts', 'home_rank', 'away_rank', 
                                'home_team_code', 'away_team_code', 
                                'home_results', 'win_rank', 'lose_rank', 'day_code'])
lsu


Wk  Pts  home_rank  away_rank  home_team_code  away_team_code  \
0  12   34         99          7             112             174   

   home_results  win_rank  lose_rank  day_code  
0             1         7         99         5

In [19]:
regr.predict(lsu)

array([24.26])

## arkansas wins
arkansas scores 39.62
lsu scores 26.12

## lsu wins
lsu 33.73
arkansas 24.26

### Results:
#### Lsu 13
#### Arkansas 10